// @anchan828/nest-sendgrid/dist/sendgrid.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var SendGridModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendGridModule = void 0;
const common_1 = require("@nestjs/common");
const sendgrid_constants_1 = require("./sendgrid.constants");
const sendgrid_service_1 = require("./sendgrid.service");
let SendGridModule = SendGridModule_1 = class SendGridModule {
    static forRoot(options) {
        if (!(options && options.apikey)) {
            throw new Error("SendGrid API Key is not defined");
        }
        return {
            module: SendGridModule_1,
            providers: [
                {
                    provide: sendgrid_constants_1.SendGridConstants.SENDGRID_MODULE_OPTIONS,
                    useValue: options,
                },
            ],
        };
    }
    static forRootAsync(options) {
        const asyncProviders = this.createAsyncProviders(options);
        return {
            module: SendGridModule_1,
            imports: [...(options.imports || [])],
            providers: [...asyncProviders],
        };
    }
    static createAsyncProviders(options) {
        if (options.useFactory) {
            return [this.createAsyncOptionsProvider(options)];
        }
        return [
            this.createAsyncOptionsProvider(options),
            {
                provide: options.useClass,
                useClass: options.useClass,
                inject: [options.inject || []],
            },
        ];
    }
    static createAsyncOptionsProvider(options) {
        if (options.useFactory) {
            return {
                provide: sendgrid_constants_1.SendGridConstants.SENDGRID_MODULE_OPTIONS,
                useFactory: options.useFactory,
                inject: options.inject || [],
            };
        }
        return {
            provide: sendgrid_constants_1.SendGridConstants.SENDGRID_MODULE_OPTIONS,
            useFactory: (optionsFactory) => __awaiter(this, void 0, void 0, function* () { return yield optionsFactory.createSendGridModuleOptions(); }),
            inject: options.useClass ? [options.useClass] : [],
        };
    }
};
SendGridModule = SendGridModule_1 = __decorate([
    common_1.Global(),
    common_1.Module({
        providers: [sendgrid_service_1.SendGridService],
        exports: [sendgrid_service_1.SendGridService],
    })
], SendGridModule);
exports.SendGridModule = SendGridModule;
